indexing-service/src/main/java/org/apache/druid/indexing/overlord/SingleTaskBackgroundRunner.java

```diff
@@ -185,6 +186,10 @@ public void stop()
       // stopGracefully for resource cleaning
       log.info("Starting graceful shutdown of task[%s].", task.getId());
       task.stopGracefully(taskConfig);
+      // Only certain tasks, primarily from unit tests, are not subclasses of AbstractTask.
+      if (task instanceof AbstractTask) {
+        ((AbstractTask) task).waitForCleanupToFinish();
+      }
```

Comment:
why bake that assumption into the code?  This class shouldn't be aware of what an `AbstractTask` is. If required, you should add a method to `Task` interface like `waitForTermination` and put the contract in the interface clearly how should that be implemented and how should it be called.

Reply:
Updated.

Reply:
Hi @abhishekagarwal87 , this PR is ready to be reviewed. Thanks.